File: web/di_core/ClickhouseConfig/Domain/SystemInfo/SystemInfo.ts
import { ClickhouseSource } from '@core/ClickhouseConfig/Domain/ClickhouseSource/ClickhouseSource';
import { RefreshBy } from '@core/ClickhouseConfig/Enum/RefreshBy/RefreshBy';
import { RefreshStatus } from '@core/ClickhouseConfig/Enum/SourceStatus/RefreshStatus';
import { SystemStatus } from '@core/ClickhouseConfig';
import { EtlJobStatus } from '@core/DataCook';

export class SystemInfo {
  constructor(
    public status: SystemStatus,
    public sources: ClickhouseSource[],
    public currentRefreshStatus: RefreshStatus,
    public createdTime: number,
    public updatedTime: number,
    public lastRefreshStatus?: RefreshStatus,
    public lastRefreshTime?: number,
    public lastRefreshBy?: RefreshBy
  ) {}

  static fromObject(obj: SystemInfo) {
    return new SystemInfo(
      obj.status,
      obj.sources,
      obj.currentRefreshStatus,
      obj.createdTime,
      obj.updatedTime,
      obj?.lastRefreshStatus,
      obj?.lastRefreshTime,
      obj?.lastRefreshBy
    );
  }

  static getIconFromRefreshStatus(status: RefreshStatus) {
    const baseUrl = 'assets/icon/data_ingestion/status';
    switch (status) {
      case RefreshStatus.Error:
        return require(`@/${baseUrl}/error.svg`);
      case RefreshStatus.Init:
        return require(`@/${baseUrl}/initialized.svg`);
      case RefreshStatus.Success:
        return require(`@/${baseUrl}/synced.svg`);
      case RefreshStatus.Running:
        return require(`@/${baseUrl}/syncing.svg`);
      default:
        return require(`@/${baseUrl}/unknown.svg`);
    }
  }

  static default() {
    return new SystemInfo(SystemStatus.Healthy, [], RefreshStatus.Init, Date.now(), Date.now());
  }
}

File: schema-service/src/main/scala/datainsider/ingestion/domain/SystemInfo.scala
package datainsider.ingestion.domain

import datainsider.ingestion.domain.RefreshBy.RefreshBy
import datainsider.ingestion.domain.RefreshStatus.RefreshStatus
import datainsider.ingestion.domain.SystemStatus.SystemStatus
import education.x.commons.Serializer
import org.apache.commons.lang3.SerializationUtils

/**
  * created 2022-07-19 4:45 PM
  *
  * @author tvc12 - Thien Vi
  */
@SerialVersionUID(20220719L)
case class SystemInfo(
    orgId: Long,
    status: SystemStatus,
    sources: Seq[ClickhouseSource],
    refreshConfig: RefreshConfig,
    currentRefreshStatus: RefreshStatus = RefreshStatus.Init,
    lastRefreshStatus: Option[RefreshStatus] = None,
    lastRefreshTime: Option[Long] = None,
    lastRefreshBy: Option[RefreshBy] = None,
    lastRefreshErrorMsg: Option[String] = None,
    createdTime: Long = System.currentTimeMillis(),
    updatedTime: Long = System.currentTimeMillis()
)

object SystemInfo {
  def default(orgId: Long): SystemInfo = {
    SystemInfo(
      orgId,
      SystemStatus.Healthy,
      sources = Seq.empty,
      refreshConfig = RefreshConfig(
        ignoredEngines = Seq.empty
      )
    )
  }

  implicit object SystemInfoSerializer extends Serializer[SystemInfo] {

    override def fromByte(bytes: Array[Byte]): SystemInfo = {
      SerializationUtils.deserialize(bytes).asInstanceOf[SystemInfo]
    }

    override def toByte(value: SystemInfo): Array[Byte] = {
      SerializationUtils.serialize(value)
    }
  }
}

object SystemStatus {
  type SystemStatus = String

  val Healthy = "healthy"
  val Unhealthy = "unhealthy"
}

object RefreshStatus {
  type RefreshStatus = String

  val Init = "init"
  val Running = "running"
  val Error = "error"
  val Success = "success"
}

object RefreshBy {
  type RefreshBy = String

  val Manual = "manual"
  val System = "system"
}
